### PR TITLE
Make signal argument to kill optional

### LINKED ISSUE #441: Not possible to call kill(pid, 0)
According to the man page:

_If sig is 0, then no signal is sent, but error checking is still performed; this can be used to check for the existence of  a process ID or process group ID._

The current signature is `pub fn kill(pid: pid_t, signal: Signal)`, and I think it should use `Option<Signal>`.

### PR DESCRIPTION
**Pull request by fiveop**
Fixes #441

### CODE REVIEW

**Comment by posborne** [comment]
I feel like a better approach might be to add a new enumerated value rather than making the signal an option.  A signal of 0 is still a signal but one that is just handled by the kernel without actually making it to the process.

I'm not sure this signal has an established name (it doesn't appear to in Linux: http://lxr.free-electrons.com/source/arch/arm/include/asm/signal.h?v=2.6.32#L31).  I suggest that we just call it `SIGZERO`.

As an added benefit, this won't break things for users using the existing API.

**Comment by fiveop** [comment]
I disagree. 0 is no signal, it represents something different. Therefore, it does not belong in the enumeration. I would not have a problem with giving it it's own name, if I not already objected to it belonging to the operation.

Since we break the API left and right at the moment, in order to get things 'nice', I do not have a problem with breaking this. It is easy to adapt code that is affected by the change.

**Comment by posborne** [comment]
Although 0 is not implemented as a signal to processes in most operating systems, I am still failing to see a strong reason to separate it.  Strace, for instance, goes so far as to name the `0` `SIG_0` in its output.

    kill(9999, SIG_0)                       = -1 ESRCH (No such process)

**Comment by posborne** [comment]
Thinking about this more, I can be OK with the change (better than not having support for this useful operation).  Moving forward, we should probably have a push for doing more documentation on our functions, particularly when there are cases like this where things might not be immediately clear from the signature.  At present, I would need to read the source (and probably go deep on the man pages) in order to figure out what passing `None` to this function does.

With some docs added, r=me

**Comment by fiveop** [comment]
I'll add a documentation comment.

File: src/sys/signal.rs [hunk 1]
```diff
@@ -388,8 +388,12 @@ pub fn pthread_sigmask(how: SigFlags,
     Errno::result(res).map(drop)
 }
 
-pub fn kill(pid: libc::pid_t, signal: Signal) -> Result<()> {
-    let res = unsafe { libc::kill(pid, signal as libc::c_int) };
+pub fn kill(pid: libc::pid_t, signal: Option<Signal>) -> Result<()> {
```
Comment (kamalmarhubi) [inline src/sys/signal.rs]: maybe we can take advantage of the [recently added](https://github.com/rust-lang/rust/pull/34828) `impl From<T> for Option<T>`. On recent enough Rust (1.12+), `kill(pid, SIGBLAH)` will still work, so it's a bit more ergonomic.

Comment (posborne) [inline src/sys/signal.rs]: Interesting, I wasn't aware of that change.  Nice!

Comment (fiveop) [inline src/sys/signal.rs]: What about us supporting Rust (1.7+), though? If someones crate wants to continue to support 1.7+, it has to change its source (which I don't have a problem with in the first place).

So asked differently: Are you saying, that we will passively take advantage of the new impl on new enough Rust versions anyway?

Comment (kamalmarhubi) [inline src/sys/signal.rs]: Yeah, that was my thought. Though perhaps we should wait until we bump minimum version to 1.12, probably in 3-6 months? Not sure how to think about it.

Comment (kamalmarhubi) [inline src/sys/signal.rs]: Trying to think about it differently: the change from `Signal` to `Option<Signal>` is breaking. The change from `Option<Signal>` to `T: Into<Option<Signal>>` is (I think?) not breaking thanks to `impl<T> From<T> for T`. (This depends on how you read "breaking" in signatures.) Our use of the `Into<Option<T>>` impl would be more akin to "progressive enhancement". A downstream crate requiring its use would have a 1.12+ minimum version, but we would not.

Comment (kamalmarhubi) [inline src/sys/signal.rs]: In any case, if making use of the new impl, we should include tests that demonstrate it works in both cases.

Comment (fiveop) [inline src/sys/signal.rs]: Sounds reasonable. And if I understand you and the matter correctly, then we could already use `T: Into<Option<Signal>>` right now.

Comment (kamalmarhubi) [inline src/sys/signal.rs]: That's my belief!

**Comment by fiveop** [comment]
@kamalmarhubi better?

**Comment by kamalmarhubi** [comment]
Could you add a test to verify that calling it with `None` works? I want to be sure there's no weird inference stuff that requires us to specify `Option::<Signal>::None` or something like that. This could even be a `no_run` doc test, I think.

With that, r=me.

**Comment by kamalmarhubi** [comment]
Also I checked and under 1.12 it allows `kill(pid, SIGKILL)`, ie no `Some`. So that's cool :-)

**Comment by fiveop** [comment]
@homu r=@kamalmarhubi
